### PR TITLE
Fix serialization/deserialization of proxy URI

### PR DESCRIPTION
**Pull request by michalpristas**
## What does this PR do?

This PR makes sure proxy URI is serialized as string as it is expected to be a string when unpacking value from config

## Why is it important?

Fixes: https://github.com/elastic/beats/issues/27187

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: libbeat/common/transport/httpcommon/proxy.go [hunk 1]
```diff
@@ -24,6 +24,9 @@ import (
 	"github.com/elastic/beats/v7/libbeat/common"
 )
 
+// ProxyURI is a URL used for proxy serialized as a string.
+type ProxyURI url.URL
```
Comment (urso) [inline libbeat/common/transport/httpcommon/proxy.go:28]: In order to make this type more reusable lets add `MarshalJSON` and unmarshallers for json, yml and config loading.

For integration with `go-ucfg` add:

```
func (p *ProxyURI) Unpack(s string) error {
  ...
}
```

File: libbeat/common/transport/httpcommon/proxy.go [hunk 2]
```diff
@@ -47,6 +50,46 @@ type HTTPClientProxySettings struct {
...
+		return nil, err
+	}
+
+	var httpHeaders http.Header
+	if len(headers) > 0 {
+		httpHeaders = http.Header{}
+		for k, v := range headers {
+			httpHeaders.Add(k, v)
+		}
+	}
```
Comment (urso) [inline libbeat/common/transport/httpcommon/proxy.go:78]: consider to introduce a custom type for headers as well.

**Comment by urso** [comment]
Please add unit tests to check the serialization and deserialization really work.

File: libbeat/common/transport/httpcommon/proxy.go [hunk 3]
```diff
@@ -47,6 +50,46 @@ type HTTPClientProxySettings struct {
...
+func (p ProxyURI) MarshalYAML() (interface{}, error) {
+	u := url.URL(p)
+	return u.String(), nil
+}
+
+// URI returns conventional url.URL structure.
+func (p ProxyURI) URI() *url.URL {
+	uri := url.URL(p)
+	return &uri
+}
```
Comment (urso) [inline libbeat/common/transport/httpcommon/proxy.go]: nit; this is just a shallow copy. How about `func (p *ProxyURI) URI() *url.URL { return (*url.URL)(p) }`

**Comment by andresrc** [comment]
Should this be backported to 7.14 also?

File: libbeat/common/transport/httpcommon/proxy.go [hunk 4]
```diff
@@ -36,17 +36,37 @@ type HTTPClientProxySettings struct {
...
+	Headers ProxyHeaders `config:"proxy_headers" yaml:"proxy_headers,omitempty"`
 
 	// Disable HTTP proxy support. Configured URLs and environment variables
 	// are ignored.
 	Disable bool `config:"proxy_disable" yaml:"proxy_disable,omitempty"`
 }
 
+// NewHTTPClientProxySettings creates a new proxy settings based on provided proxy information.
+func NewHTTPClientProxySettings(url string, headers map[string]string, disable bool) (*HTTPClientProxySettings, error) {
+	proxyURI, err := common.ParseURL(url)
```
Comment (urso) [inline libbeat/common/transport/httpcommon/proxy.go:52]: use `NewProxyURIFromString`?

File: libbeat/common/transport/httpcommon/proxy_uri.go [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+
+func NewProxyURIFromURL(u url.URL) *ProxyURI {
+	p := ProxyURI(u)
+	return &p
+}
+
+// MarshalYAML serializes URI as a string.
+func (p ProxyURI) MarshalYAML() (interface{}, error) {
+	u := url.URL(p)
+	return u.String(), nil
```
Comment (urso) [inline libbeat/common/transport/httpcommon/proxy_uri.go:48]: we should implement `Stringer` for `ProxyURI` right away.

File: libbeat/common/transport/httpcommon/proxy_uri.go [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+}
+
+// MarshalJSON serializes URI as a string.
+func (p ProxyURI) MarshalJSON() ([]byte, error) {
+	u := url.URL(p)
+	fmt.Println(u.String())
+	return []byte(fmt.Sprintf("\"%s\"", u.String())), nil
```
Comment (urso) [inline libbeat/common/transport/httpcommon/proxy_uri.go:55]: better use `json.Marshal(u.String())`. Make sure that the URI string is properly quoted/escaped.

File: libbeat/common/transport/httpcommon/proxy_uri.go [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+	return NewProxyURIFromURL(*u), nil
+}
+
+func NewProxyURIFromURL(u url.URL) *ProxyURI {
+	p := ProxyURI(u)
+	return &p
+}
+
+// MarshalYAML serializes URI as a string.
+func (p ProxyURI) MarshalYAML() (interface{}, error) {
```
Comment (urso) [inline libbeat/common/transport/httpcommon/proxy_uri.go:46]: `ProxyURI` is stored by pointer in the config struct. Will the yaml marshaller find that method? Reflection can be a pain at times, reflection rules are not as forgiving about pointers vs. non-pointers as the compiler. Consider to use pointers more consistently.

**Comment by urso** [comment]
The changes here also affect how `HTTPClientProxySettings` is serialized/deserialized. Please add tests for `HTTPClientProxySettings` as well.

File: libbeat/common/transport/httpcommon/proxy_uri.go [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+	*p = *uri
+	return nil
+}
+
+// URI returns conventional url.URL structure.
+func (p *ProxyURI) URI() *url.URL {
+	return (*url.URL)(p)
+}
+
+// MarshalJSON serializes URI as a string.
```
Comment (urso) [inline libbeat/common/transport/httpcommon/proxy_uri.go:98]: godoc

File: libbeat/esleg/eslegclient/connection.go [hunk 9]
```diff
@@ -167,7 +167,7 @@ func NewClients(cfg *common.Config) ([]Connection, error) {
 	}
 
 	if proxyURL := config.Transport.Proxy.URL; proxyURL != nil {
-		logp.Info("using proxy URL: %s", proxyURL)
+		logp.Info("using proxy URL: %s", proxyURL.URI().String())
```
Comment (urso) [inline libbeat/esleg/eslegclient/connection.go:170]: we already implement `String` on proxyURL. No need to change that line.

**Review by urso — APPROVED** [review]

**Pull request merged**